swap/peer.go

```diff
@@ -30,18 +34,142 @@ var ErrDontOwe = errors.New("no negative balance")
 // Peer is a devp2p peer for the Swap protocol
 type Peer struct {
 	*protocols.Peer
+	lock               sync.RWMutex
 	swap               *Swap
 	beneficiary        common.Address
 	contractAddress    common.Address
 	lastReceivedCheque *Cheque
+	lastSentCheque     *Cheque
+	balance            int64
 }
 
 // NewPeer creates a new swap Peer instance
-func NewPeer(p *protocols.Peer, s *Swap, beneficiary common.Address, contractAddress common.Address) *Peer {
-	return &Peer{
+func NewPeer(p *protocols.Peer, s *Swap, beneficiary common.Address, contractAddress common.Address) (*Peer, error) {
+	peer := &Peer{
 		Peer:            p,
 		swap:            s,
 		beneficiary:     beneficiary,
 		contractAddress: contractAddress,
 	}
+
+	var err error
+	if peer.lastReceivedCheque, err = s.loadLastReceivedCheque(p.ID()); err != nil {
+		return nil, err
+	}
+
+	if peer.lastSentCheque, err = s.loadLastSentCheque(p.ID()); err != nil {
+		return nil, err
+	}
+
+	if peer.balance, err = s.loadBalance(p.ID()); err != nil {
+		return nil, err
+	}
+	return peer, nil
+}
+
+func (p *Peer) getLastReceivedCheque() *Cheque {
+	return p.lastReceivedCheque
+}
+
+func (p *Peer) getLastSentCheque() *Cheque {
+	return p.lastSentCheque
+}
+
+func (p *Peer) setLastReceivedCheque(cheque *Cheque) error {
+	p.lastReceivedCheque = cheque
+	return p.swap.saveLastReceivedCheque(p.ID(), cheque)
+}
+
+func (p *Peer) setLastSentCheque(cheque *Cheque) error {
+	p.lastSentCheque = cheque
+	return p.swap.saveLastSentCheque(p.ID(), cheque)
+}
+
+func (p *Peer) getLastCumulativePayout() uint64 {
+	lastCheque := p.getLastReceivedCheque()
+	if lastCheque != nil {
+		return lastCheque.CumulativePayout
+	}
+	return 0
+}
+
+func (p *Peer) setBalance(balance int64) error {
+	p.balance = balance
+	return p.swap.saveBalance(p.ID(), balance)
```

Comment:
As balance is now defined on the `peer`, wouldn't it be better to have the functions related to `balance`, also in the peer? 
That way, you could just do `p.saveBalance`.
The same holds true for `saveLastSentCheque`

Reply:
I prefer to keep it as it is. `Swap` owns the store so it is also fitting for functions that write to the store to be there. Moving this to peer would mean either accessing `p.swap.store` directly or adding an unnecessary indirection.